sway/desktop/render.c

```diff
@@ -186,6 +401,52 @@ static void render_surface_iterator(struct sway_output *output,
 	scale_box(&dst_box, wlr_output->scale);
 
 	data->deco_data.corner_radius *= wlr_output->scale;
+
+	// Render blur
+	struct decoration_data deco_data = data->deco_data;
+	bool opaque = true;
+	if (view && view->surface) {
+		struct wlr_gles2_texture_attribs attribs;
+		wlr_gles2_texture_get_attribs(view->surface->buffer->texture, &attribs);
+		opaque = (view->surface->opaque || !attribs.has_alpha);
+	}
+	if (view
+			// Don't render blur for subsurfaces
+			&& view->surface == surface
+			&& deco_data.blur
+			&& view->container->blur_enabled
+			// Check if window has alpha
+			&& !(opaque && view->container->alpha >= 1.0f && deco_data.alpha >= 1.0f)
+			&& should_parameters_blur()) {
+		int width, height;
+		wlr_output_transformed_resolution(output->wlr_output, &width, &height);
+		struct wlr_fbox src_box = {
+			.x = 0,
+			.y = 0,
+			.width = width,
+			.height = height,
+		};
+
+		// Use an empty opaque region if opacity is applied to the whole window
+		// through swayipc
+		pixman_region32_t opaque_region;
+		pixman_region32_init(&opaque_region);
+		if (view->container->alpha < 1.0f || deco_data.alpha < 1.0f) {
+			pixman_region32_union_rect(&opaque_region, &opaque_region, 0, 0, 0, 0);
+		} else {
+			pixman_region32_copy(&opaque_region, &surface->opaque_region);
+		}
+
+		// Use the pre-rendered blurred buffer for tiled windows due tiled
+		// that all tiled windows are displaying the same content under the windows
+		bool is_floating = container_is_floating(view->container);
+		render_blur(!is_floating, output, output_damage, &src_box, &dst_box,
+				&opaque_region, surface->current.width, surface->current.height,
+				surface->current.scale, deco_data, config->blur_params);
+
+		pixman_region32_fini(&opaque_region);
+	}
```

Comment:
would this be better suited to live in render_view? Right now it seems like we just want to blur views anyway

Reply:
That would be nice! Before when it was located there, resizing a window would make the blur flicker (maybe due to the view being null?)

Reply:
That would be nice! Before when it was located there, resizing a window would make the blur flicker (maybe due to the view being null?)

Reply:
Hmm, good reminder about the flicker, something to look into with a `null` check